src/tsolvers/lrasolver/LRASolver.cc

```diff
@@ -96,225 +96,28 @@ lbool LRASolver::getPolaritySuggestion(PTRef ptref) const {
+        else if (itpAlgorithm == itp_lra_alg_factor) { return FarkasItpOptions::useFlexibleFarkasAlgorithm(opensmt::Real(config.getLRAStrengthFactor())); }
+        else if (itpAlgorithm == itp_lra_alg_decomposing_strong) { return FarkasItpOptions::useDecomposingFarkasAlgorithm(); }
+        else if (itpAlgorithm == itp_lra_alg_decomposing_weak) { return FarkasItpOptions::useDualDecomposingFarkasAlgorithm(); }
+        else { // SHOULD NOT HAPPEN!
             assert(false);
+            return FarkasItpOptions::useFarkasAlgorithm();
         }
-        assert( color == I_A
-                || color == I_AB
-                || color == I_B );
-
-        PTRef exp_pt = explanation[i].tr;
-        if(labels != nullptr && labels->find(exp_pt) != labels->end())
-        {
-            if(color != I_UNDEF){
-                // Partitioning and labels can disagree under conditions that according to partitioning
-                // the explanation can be in both parts, but the label can be more strict
-//                std::cerr << "Color disagreement for term: " << logic.printTerm(explanation[i].tr) << '\n';
-//                std::cerr << "Color from partitioning: " << color << '\n';
-//                std::cerr << "Color from labels: " << labels->find(exp_pt)->second << '\n';
-                assert(color == I_AB || color == labels->find(exp_pt)->second);
-            }
-            // labels have priority of simple partitioning information
-            color = labels->find(exp_pt)->second;
-            //cout << "; PTRef " << logic.printTerm(exp_pt) << " has Boolean color " << color << endl;
-        }
-
-        //assert( color == I_A || color == I_B );
-
-        // Add the conflict to the interpolant (multiplied by the coefficient)
-        if(color == I_A || color == I_AB)
-        {
-            if (explanation[i].sgn == l_False)
-            {
-                interpolant.addExprWithCoeff(LAExpression(logic, explanation[i].tr, false), explanationCoefficients[i]);
-                delta_flag=true;
-            }
-            else
-            {
-                interpolant.addExprWithCoeff(LAExpression(logic, explanation[i].tr, false), -explanationCoefficients[i]);
-            }
-        }
-        if(color == I_B || color == I_AB)
-        {
-            if (explanation[i].sgn == l_False)
-            {
-                interpolant_dual.addExprWithCoeff(LAExpression(logic, explanation[i].tr, false), explanationCoefficients[i]);
-                // TODO: investigate where delta_flag_dual should be used and how it should be used properly
-                delta_flag_dual=true;
-            }
-            else
-            {
-                interpolant_dual.addExprWithCoeff(LAExpression(logic, explanation[i].tr, false), -explanationCoefficients[i]);
-            }
-        }
-    }
+    }();
+    return interpolator.getInterpolant(options);
```

Comment:
This function has a similar check to the above lambda.  Maybe with refactoring as I commented above on splitting the long function, or alternatively with a factory approach that returns an object with a derived type, one could avoid repeating the code?